backend/fms_core/services/library.py

```diff
@@ -184,4 +192,86 @@ def update_library(sample, **kwargs):
         except Exception as e:
             errors.append(str(e))
 
-    return sample, errors, warnings
+    return derived_sample, errors, warnings
+
+
+def convert_library_concentration_from_nm_to_ngbyul(source_sample, concentration_nm):
+    """
+                 Converts the concentration of a library from nM to ng/uL by calculating each individual concentration
+                 of the derived samples and adjusting it with their respective volume ratio.
+
+                 Args:
+                     `sample_source`: The source library to be converted.
+                     `concentration_nm`: The library concentration in nM.
+
+                 Returns:
+                     The resulting concentration in ng/uL
+            """
+    errors = []
+    warnings = []
+
+    if source_sample is None:
+        errors.append(f'Missing sample')
+
+    final_concentration = 0
+    for derived_sample in source_sample.derived_samples.all():
+        # Compute the size of each library and its volume ratio
+        library = derived_sample.library
+        volume_ratio = DerivedBySample.objects.get(derived_sample=derived_sample, sample=source_sample).volume_ratio
+        if library.library_size:
+            # Convert the concentration
+            individual_concentration = convert_concentration_from_nm_to_ngbyul(concentration_nm,
+                                                                               library.molecular_weight_approx,
+                                                                               library.library_size)
+            if individual_concentration is None:
+                errors.append(f'Failed to convert the concentration of this library {derived_sample.biosample.alias}.')
```

Comment:
I don't like too much receiving a sample as param and returning an error from the derived_sample. Since this returns right away you can possibly mention the sample name instead of the Alias. The failure is for the whole sample/library.